sql: support prepare for CASE statements

madelynnblue opened the pull request:
Also add protecting against other statements where valargs are not fully
supported. Say that $2 gets inferred but $1 does not. The previous code
would panic because the slice isn't long enough (expected to be of length
2 when it indexed into pq.inTypes[1]) because it only made pq.inTypes
length 1 since that's how many valargs it inferred.

Fixes #4024

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4142)

<!-- Reviewable:end -->

tamird commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/parser/type_check.go, line 88 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hsXHSYzh7SyGuFy3V:-K9hsXHSYzh7SyGuFy3W:-552019735) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/parser/type_check.go#L88)):*
nit: valArgType

---

*[sql/pgwire/v3.go, line 338 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hsjr8TZY3jwq-Kzwh:-K9hsjr8TZY3jwq-Kzwi:310388902) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire/v3.go#L338)):*
add a comment explaining (what i assume is) the one-indexing of this

---

*[sql/pgwire/v3.go, line 340 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hsaCf0asOUZC7ACbG:-K9hsaCf0asOUZC7ACbH:-1673153643) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire/v3.go#L340)):*
can you log some additional information here? if not, remove the Sprinft call.

---

*[sql/pgwire_test.go, line 174 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hssoqZ6Clh7D24ylm:-K9hssoqZ6Clh7D24yln:338896565) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire_test.go#L174)):*
we should use angle brackets like the other similar errors.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: support prepare for CASE statements

Also add protecting against other statements where valargs are not fully
supported. Say that $2 gets inferred but $1 does not. The previous code
would panic because the slice isn't long enough (expected to be of length
2 when it indexed into pq.inTypes[1]) because it only made pq.inTypes
length 1 since that's how many valargs it inferred.

Fixes #4024

mjibson commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/parser/type_check.go, line 88 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hsXHSYzh7SyGuFy3V:-K9hu_oYWHXvrarOvaPQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/parser/type_check.go#L88)):*
Done.

---

*[sql/pgwire/v3.go, line 338 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hsjr8TZY3jwq-Kzwh:-K9huqsczlk62kqhhgb6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire/v3.go#L338)):*
Done.

---

*[sql/pgwire/v3.go, line 340 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hsaCf0asOUZC7ACbG:-K9hv7rgpm2SPAW2am4D:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire/v3.go#L340)):*
Done.

---

*[sql/pgwire_test.go, line 174 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hssoqZ6Clh7D24ylm:-K9hvClIvkGQsAPALmK8:1960553789) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire_test.go#L174)):*
See `SELECT $1` above, which doesn't use angle brackets. I think if we want to make this change we should do them all at once and in another PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/pgwire/v3.go, line 340 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hsaCf0asOUZC7ACbG:-K9hy9-75IqPmS-kvCVT:450800964) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire/v3.go#L340)):*
nit: valArgs

---

*[sql/pgwire_test.go, line 174 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hssoqZ6Clh7D24ylm:-K9hyCqP_Iu8gIB7M6xJ:-2139350909) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire_test.go#L174)):*
Sounds good. Want to file an issue? Be sure to mark it easy.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-:-K9hyFjdv2XGN--WhSLw:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/pgwire_test.go, line 174 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142#-K9hssoqZ6Clh7D24ylm:-K9i0AUd8x_fnINA_loK:-1004562899) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0153fa1ddaff5aeb9c1c85ffe26117e5f7ce8d0/sql/pgwire_test.go#L174)):*
If there is no strong reason to do it one way or the other, I would try to format it similarly to the errors postgres would return.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4142)*
<!-- Sent from Reviewable.io -->
